### ISSUE TITLE
periodic padding, convolutional neural net

### ISSUE THREAD

**Issue by mgbukov** [google/jax#1393]
Does jax allow one to implement convolutional and pooling layers with periodic padding?

A trivial (but probably slow) solution would be to manually preprocess the input data at every layer using `jax.numpy.pad(mode='wrap')` and use this in combination with `padding='VALID'` in the conv layer, but this has to be done separately every time and it differs depending on stride and filter size. 

Are there any other alternatives?

**Comment by mattjj** [comment]
Thanks for the question!

My understanding is there are two ways to read this:
1. can JAX expose existing fast kernels for doing conv/pooling with periodic padding (which would be faster than doing the `np.pad` manually)?
2. can we add periodic padding as a JAX library feature, even if under-the-hood it effectively calls `np.pad`, since that would be convenient?

For 1, my reading of [XLA's `ConvGeneralDilated` docs](https://www.tensorflow.org/xla/operation_semantics#convwithgeneralpadding_convolution) is that there is no option for periodic padding ("The padding argument specifies the amount of zero padding to be applied to the base area"). @hawkinsp will correct me if I'm wrong. Do you know if cuDNN has kernels for this kind of thing? What do TF / PyTorch do?

For 2, that could be added. We could mark this as a JAX feature request, and probably also add "good first issue". Pull requests welcome ;)

**Comment by mgbukov** [comment]
For 2, I looked into adding such a feature. Half-way through, I noticed that `jit`-ting the code leads to a slowdown, which is weird. 

The `PeriodicConv` layer is a mere copy of the master branch `jax.experimental.stax.GeneralConv()` with a few extra lines added (see `DNN_architectures.txt`).

To reproduce the slowdown, run  `test_CNN.txt` (attached). I am not sure that the slowdown is related to the periodic padding feature. 

It looks like the CNN code is just being slowed down by adding `jit`. If I remove the `jit` from lines 66, 82 and 84 in `test_CNN.txt`, the code speeds up. Any ideas where this might be coming from? 

For comparison, I also tried a single dense layer (toggle in line 21 in `text_CNN.txt`): for the dense layer, `jit`-ting leads to a speedup as expected. I use the same nonlinearity in both cases with the same number of NN parameters. 

[DNN_architectures.txt](https://github.com/google/jax/files/3678763/DNN_architectures.txt)
[test_CNN.txt](https://github.com/google/jax/files/3678764/test_CNN.txt)

**Comment by mgbukov** [comment]
I looked more closely: it seems like the periodic padding is unrelated to the slow compile issue. I will continue with case 2, but comments are welcome. 

The slow compile happens in computing the per-sample gradients. Once the code is compiled, it actually runs really fast. Any ideas what causes this behavior, and the related message:
```
Very slow compile?  If you want to file a bug, run with envvar XLA_FLAGS=--xla_dump_to=/tmp/foo and attach the results.
```
are of welcome.

**Comment by PWhiddy** [comment]
Any update on this?

**Comment by mattjj** [comment]
I think this thread is pretty stale, so I'm closing it. Don't hesitate to open new issues if you have new versions of this question!
